### PR TITLE
[qr-code] Align QR Code format with the spec

### PR DESCRIPTION
**Pull request by Damian-Nordic**
#### Problem
QR Code format is no longer aligned with the spec after recent changes.

#### Change overview
1. Replace boolean requiresCustomFlow with 2-bit field with
   the following meaning:
   0 - device automatically enters paring mode upon power-up
   1 - device needs user interaction to enter pairing mode
   2 - device uses a vendor-specific flow that should be
       retrieved from the distributed ledger.
2. Replace CH: prefix with MT:

#### Testing
* There are many unit tests that cover QR Code and Manual Pairing Code parsing.
  * Reviewed existing test cases and added a few more
* Additionally, tested the change manually using nRF Connect lighting-app and Android CHIP Tool
* Also, tested parsing QR Codes with optional data using Python CHIP Controller `setup-payload parse-qr` command.

### CODE REVIEW

**Review by msandstedt — APPROVED** [review]

File: examples/lock-app/cc13x2x7_26x2x7/README.md [hunk 1]
```diff
@@ -226,15 +226,15 @@ This is done by scanning a QR code. A URL will be displayed on the lock-app's
 log ([UART terminal](#viewing-logging-output)). It will look like the following:
 
 ```
-SetupQRCode:  [CH:.81TM -00 0C9SS0]
+SetupQRCode:  [MT:.81TM -00 0C9SS0]
```
Comment (bzbarsky-apple) [inline examples/lock-app/cc13x2x7_26x2x7/README.md:229]: I expect the value here, and others, are wrong since we switched to base-39.  Followup issue to fix them?

File: src/darwin/Framework/CHIPTests/CHIPSetupPayloadParserTests.m [hunk 2]
```diff
@@ -109,15 +109,15 @@ - (void)testOnboardingPayloadParser_QRCode_NoError
     XCTAssertEqual(payload.setUpPINCode.unsignedIntegerValue, 2048);
     XCTAssertEqual(payload.vendorID.unsignedIntegerValue, 12);
     XCTAssertEqual(payload.productID.unsignedIntegerValue, 1);
-    XCTAssertFalse(payload.requiresCustomFlow);
+    XCTAssertEqual(payload.commissioningFlow, kCommissioningFlowStandard);
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIPTests/CHIPSetupPayloadParserTests.m:112]: Why the payload change from custom to standard?

File: src/darwin/Framework/CHIPTests/CHIPSetupPayloadParserTests.m [hunk 3]
```diff
@@ -140,7 +140,7 @@ - (void)testOnboardingPayloadParser_NFC_NoError
...
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIPTests/CHIPSetupPayloadParserTests.m:143]: Again, why the change from custom to standard?

File: src/darwin/Framework/CHIPTests/CHIPSetupPayloadParserTests.m [hunk 4]
```diff
@@ -171,7 +171,7 @@ - (void)testManualParser
     XCTAssertEqual(payload.setUpPINCode.unsignedIntegerValue, 123456780);
     XCTAssertEqual(payload.vendorID.unsignedIntegerValue, 1);
     XCTAssertEqual(payload.productID.unsignedIntegerValue, 1);
-    XCTAssertTrue(payload.requiresCustomFlow);
+    XCTAssertEqual(payload.commissioningFlow, kCommissioningFlowCustom);
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIPTests/CHIPSetupPayloadParserTests.m:174]: And here, why the change?

File: src/darwin/Framework/CHIPTests/CHIPSetupPayloadParserTests.m [hunk 5]
```diff
@@ -190,7 +190,7 @@ - (void)testQRCodeParser_Error
 {
     NSError * error;
     CHIPQRCodeSetupPayloadParser * parser =
-        [[CHIPQRCodeSetupPayloadParser alloc] initWithBase38Representation:@"CH:J5L900CK70WWI0000"];
+        [[CHIPQRCodeSetupPayloadParser alloc] initWithBase38Representation:@"NT:R5L90MP500K64J00000"];
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIPTests/CHIPSetupPayloadParserTests.m:193]: Why the change from invalid code to invalid URL prefix here?  Should keep testing invalid code, I'd think....

File: src/darwin/Framework/CHIPTests/CHIPSetupPayloadParserTests.m [hunk 6]
```diff
@@ -211,16 +211,16 @@ - (void)testQRCodeParser
     XCTAssertEqual(payload.setUpPINCode.unsignedIntegerValue, 2048);
     XCTAssertEqual(payload.vendorID.unsignedIntegerValue, 12);
     XCTAssertEqual(payload.productID.unsignedIntegerValue, 1);
-    XCTAssertFalse(payload.requiresCustomFlow);
+    XCTAssertEqual(payload.commissioningFlow, kCommissioningFlowStandard);
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIPTests/CHIPSetupPayloadParserTests.m:214]: Why the change to standard?

File: src/darwin/Framework/CHIP/CHIPSetupPayload_Internal.h [hunk 7]
```diff
@@ -17,7 +17,8 @@
 
 #ifdef __cplusplus
 - (id)initWithSetupPayload:(chip::SetupPayload)setupPayload;
-- (CHIPRendezvousInformationFlags)valueOf:(chip::RendezvousInformationFlags)value;
+- (CHIPRendezvousInformationFlags)convertRendezvousFlags:(chip::RendezvousInformationFlags)value
+- (CHIPCommissioningFlow)convertCommissioningFlow:(chip::CommissioningFlow)value
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/CHIPSetupPayload_Internal.h:21]:
```suggestion
- (CHIPRendezvousInformationFlags)convertRendezvousFlags:(chip::RendezvousInformationFlags)value;
- (CHIPCommissioningFlow)convertCommissioningFlow:(chip::CommissioningFlow)value;
```
might help the Darwin build.

File: src/darwin/Framework/CHIPTests/CHIPSetupPayloadParserTests.m [hunk 8]
```diff
@@ -231,7 +231,7 @@ - (void)testQRCodeParserWithOptionalData
...
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIPTests/CHIPSetupPayloadParserTests.m:234]: Why the change to standard?

File: src/setup_payload/QRCodeSetupPayloadParser.cpp [hunk 9]
```diff
@@ -376,13 +376,14 @@ CHIP_ERROR QRCodeSetupPayloadParser::populatePayload(SetupPayload & outPayload)
     static_assert(kProductIDFieldLengthInBits <= 16, "Won't fit in uint16_t");
     outPayload.productID = static_cast<uint16_t>(dest);
 
-    err = readBits(buf, indexToReadFrom, dest, kCustomFlowRequiredFieldLengthInBits);
+    err = readBits(buf, indexToReadFrom, dest, kCommissioningFlowFieldLengthInBits);
     SuccessOrExit(err);
-    static_assert(kCustomFlowRequiredFieldLengthInBits <= 8, "Won't fit in uint8_t");
-    outPayload.requiresCustomFlow = static_cast<uint8_t>(dest);
+    static_assert(kCommissioningFlowFieldLengthInBits <= 8, "Won't fit in uint8_t");
+    outPayload.commissioningFlow = static_cast<CommissioningFlow>(dest);
```
Comment (bzbarsky-apple) [inline src/setup_payload/QRCodeSetupPayloadParser.cpp:382]: Shouldn't we now be asserting that `CommissioningFlow` has enough bits to fit this?  Where does `uint8_t` come in?

File: src/darwin/Framework/CHIPTests/CHIPSetupPayloadParserTests.m (hunk 2 above)
Comment (Damian-Nordic) [reply src/darwin/Framework/CHIPTests/CHIPSetupPayloadParserTests.m:112]: I changed to the standard whenever it was `XCTAssertFalse` before.

File: src/darwin/Framework/CHIPTests/CHIPSetupPayloadParserTests.m (hunk 3 above)
Comment (Damian-Nordic) [reply src/darwin/Framework/CHIPTests/CHIPSetupPayloadParserTests.m:143]: I changed to the standard whenever it was `XCTAssertFalse` before.

File: src/darwin/Framework/CHIPTests/CHIPSetupPayloadParserTests.m (hunk 4 above)
Comment (Damian-Nordic) [reply src/darwin/Framework/CHIPTests/CHIPSetupPayloadParserTests.m:174]: The boolean changed to enum, so I had to change it. Or am I missing something?

File: src/darwin/Framework/CHIPTests/CHIPSetupPayloadParserTests.m (hunk 5 above)
Comment (Damian-Nordic) [reply src/darwin/Framework/CHIPTests/CHIPSetupPayloadParserTests.m:193]: I wasn't sure what the test case was supposed to test. I can switch to a code which results in the "invalid argument" error though.

File: src/setup_payload/QRCodeSetupPayloadParser.cpp (hunk 9 above)
Comment (Damian-Nordic) [reply src/setup_payload/QRCodeSetupPayloadParser.cpp:382]: outPayload.commissioningFlow has 8 bits. I will update the static assert message though.

File: src/darwin/Framework/CHIPTests/CHIPSetupPayloadParserTests.m (hunk 5 above)
Comment (Damian-Nordic) [reply src/darwin/Framework/CHIPTests/CHIPSetupPayloadParserTests.m]: Done, by the way, while trying to find a candidate for the test case I found that we silently pass some invalid QR codes, so fixed that and added more test cases.

**Review by andy31415 — APPROVED** [review]

**Comment by woody-apple** [comment]
@shana-apple ?

File: src/darwin/Framework/CHIPTests/CHIPSetupPayloadParserTests.m (hunk 3 above)
Comment (bzbarsky-apple) [reply src/darwin/Framework/CHIPTests/CHIPSetupPayloadParserTests.m:143]: Ah, I missed the "False".  Sorry...

File: src/darwin/Framework/CHIPTests/CHIPSetupPayloadParserTests.m (hunk 4 above)
Comment (bzbarsky-apple) [reply src/darwin/Framework/CHIPTests/CHIPSetupPayloadParserTests.m:174]: No, this was me being confused, I think.

File: src/setup_payload/QRCodeSetupPayloadParser.cpp [hunk 10]
```diff
@@ -376,13 +376,15 @@ CHIP_ERROR QRCodeSetupPayloadParser::populatePayload(SetupPayload & outPayload)
     static_assert(kProductIDFieldLengthInBits <= 16, "Won't fit in uint16_t");
     outPayload.productID = static_cast<uint16_t>(dest);
 
-    err = readBits(buf, indexToReadFrom, dest, kCustomFlowRequiredFieldLengthInBits);
+    err = readBits(buf, indexToReadFrom, dest, kCommissioningFlowFieldLengthInBits);
     SuccessOrExit(err);
-    static_assert(kCustomFlowRequiredFieldLengthInBits <= 8, "Won't fit in uint8_t");
-    outPayload.requiresCustomFlow = static_cast<uint8_t>(dest);
+    static_assert(kCommissioningFlowFieldLengthInBits <= 8 * sizeof(CommissioningFlow), "Won't fit in CommissioningFlow");
```
Comment (bzbarsky-apple) [inline src/setup_payload/QRCodeSetupPayloadParser.cpp:381]: Perhaps
```suggestion
    static_assert(kCommissioningFlowFieldLengthInBits <= std::numeric_limits<std::underlying_type_t<CommissioningFlow>>::digits, "Won't fit in CommissioningFlow");
```
to be very foolproof.

**Review by bzbarsky-apple — APPROVED** [review]

File: src/darwin/Framework/CHIP/CHIPSetupPayload.h [hunk 11]
```diff
@@ -28,6 +28,13 @@ typedef NS_ENUM(NSUInteger, CHIPRendezvousInformationFlags) {
     kRendezvousInformationAllMask = kRendezvousInformationSoftAP | kRendezvousInformationBLE | kRendezvousInformationOnNetwork,
 };
 
+typedef NS_ENUM(NSUInteger, CHIPCommissioningFlow) {
+    kCommissioningFlowStandard = 0, // Device automatically enters pairing mode upon power-up
+    kCommissioningFlowUserActionRequired = 1, // Device requires a user interaction to enter pairing mode
+    kCommissioningFlowCustom = 2, // Commissioning steps should be retrieved from the distributed compliance ledger
+    kCommissioningFlowInvalid = 3,
```
Comment (shana-apple) [inline src/darwin/Framework/CHIP/CHIPSetupPayload.h:35]: Does the spec mention invalid enum, I thought it was just a reserved bit. 
I'd expect the code to have 3 enums and if we ever get value 3, we'd return an error. Or the adopter would know this is invalid as we did not define it.

Comment (Damian-Nordic) [reply src/darwin/Framework/CHIP/CHIPSetupPayload.h:35]: It is a reserved value, but here I need something to represent it while converting to the Darwin type. Normally it shouldn't happen.

**Review by LuDuda — APPROVED** [review]
